xds/src/main/java/io/grpc/xds/Matcher.java

```diff
@@ -0,0 +1,476 @@
+    }
+
+    static HeaderMatcher forPresent(String name, boolean present, boolean inverted) {
+      checkNotNull(name, "name");
+      return HeaderMatcher.create(name, null, null, null, present, null, null, inverted);
+    }
+
+    static HeaderMatcher forPrefix(String name, String prefix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(prefix, "prefix");
+      return HeaderMatcher.create(name, null, null, null, null, prefix, null, inverted);
+    }
+
+    static HeaderMatcher forSuffix(String name, String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      checkNotNull(suffix, "suffix");
+      return HeaderMatcher.create(name, null, null, null, null, null, suffix, inverted);
+    }
+
+    private static HeaderMatcher create(String name, @Nullable String exactValue,
+        @Nullable Pattern safeRegEx, @Nullable Range range,
+        @Nullable Boolean present, @Nullable String prefix,
+        @Nullable String suffix, boolean inverted) {
+      checkNotNull(name, "name");
+      return new AutoValue_Matcher_HeaderMatcher(name, exactValue, safeRegEx, range, present,
+          prefix, suffix, inverted);
+    }
+
+    /** Represents an integer range. */
+    @AutoValue
+    abstract static class Range {
+      abstract long start();
+
+      abstract long end();
+
+      static Range create(long start, long end) {
+        return new AutoValue_Matcher_HeaderMatcher_Range(start, end);
+      }
+    }
+
+    @Override
+    public boolean matches(EvaluateArgs args) {
+      String value = args.getHeaderValue(name());
+      if (present() != null) {
+        return (value == null) == present().equals(inverted());
+      }
+      if (value == null) {
+        return false;
+      }
+      boolean baseMatch;
+      if (exactValue() != null) {
+        baseMatch = exactValue().equals(value);
+      } else if (safeRegEx() != null) {
+        baseMatch = safeRegEx().matches(value);
+      } else if (range() != null) {
+        long numValue;
+        try {
+          numValue = Long.parseLong(value);
+          baseMatch = numValue >= range().start()
+              && numValue <= range().end();
+        } catch (NumberFormatException ignored) {
+          baseMatch = false;
+        }
+      } else if (prefix() != null) {
+        baseMatch = value.startsWith(prefix());
+      } else {
+        baseMatch = value.endsWith(suffix());
+      }
+      return baseMatch != inverted();
+    }
+  }
+
+  /** Represents a fractional value. */
+  @AutoValue
+  abstract static class FractionMatcher {
+    abstract int numerator();
+
+    abstract int denominator();
+
+    static FractionMatcher create(int numerator, int denominator) {
+      return new AutoValue_Matcher_FractionMatcher(numerator, denominator);
+    }
+  }
+
+  /** Represents various ways to match a string .*/
+  @AutoValue
+  abstract static class StringMatcher extends Matcher {
```

Comment:
Such a class hierarchy is very weird. If `boolean matches(EvaluateArgs args)` does not work for StringMatcher, then it is not a Matcher, it should not extend Matcher.